Add At/Index class (#43)

leighman opened the pull request:
Thought I'd give this a try - I think I've understood correctly...
Added At for StrMap and Set based on the monocle module. Noticed that PureScript has `set` always with a `Maybe` - do we want that here instead?
Is it worth adding `atRecord` or `atOption` as PS does?
Wasn't sure about the composition stuff, would appreciate some guidance on that, then could probably look at doing `Index` too.

gcanti commented on src/index.ts:
```diff
@@ -568,6 +571,21 @@ export class Traversal<S, A> {
+  return new At(at => new Lens(s => SM.lookup(at, s), a => s => a.fold(SM.remove(at, s), x => SM.insert(at, x, s))))
+}
+
+export function atSet<A = never>(setoid: Setoid<A>): At<Set<A>, A, boolean> {
```
I think that `S.member(setoid)`, `S.insert(setoid)` and `S.remove(setoid)` can be memoized

```ts
export function atSet<A = never>(setoid: Setoid<A>): At<Set<A>, A, boolean> {
  const member = S.member(setoid)
  const insert = S.insert(setoid)
  const remove = S.remove(setoid)
  return new At(
    at => new Lens(s => member(s)(at), a => s => (a ? insert(at, s) : remove(at, s)))
  )
}
```

gcanti commented:
Thanks @leighman

> Noticed that PureScript has set always with a Maybe - do we want that here instead?

Since this library is a porting of Scala's monocle I'd stick with monocle definitions

> Added At for StrMap and Set based on the monocle module

Let's move them in specific files so that users are not forced to load unwanted dependencies, something along the lines of

```
- src
  - index.ts (<= At definition)
  - At (<= At instances)
    - StrMap.ts (<= atStrMap instance)
    - Set.ts (<= atSet instance)
```

> Is it worth adding atRecord or atOption as PS does?

- atOption: no
- atRecord: not sure what's this, could you please elaborate?

> Wasn't sure about the composition stuff

What do you mean, for example `fromIso`?

gcanti commented:
> I guess they could write one if needed

Yes, if we add a `fromIso` method to `At`

```ts
export class At<S, I, A> {
  ...
  /** lift an instance of `At` using an `Iso` */
  fromIso<T>(iso: Iso<T, S>): At<T, I, A> {
    return new At(i => iso.composeLens(this.at(i)))
  }
}
```

then people can build an `At` for `Record<string, A>` pretty easily

```ts
const isoStrMap = <A = never>() => new Iso<Record<string, A>, SM.StrMap<A>>(s => new SM.StrMap(s), a => a.value)

const atRecord = atStrMap<string>().fromIso(isoStrMap())
```

leighman commented:
Added `Index` too.
I used `strmapIndex` and `arrayIndex` because that's consistent with monocle but it's `atStrMap` in `At`. Should we unify them?

leighman commented:
Renamed.

> Are src/At/index.ts, src/Index/index.ts useful? Looks like they are just a burden to maintain

I agree. Misunderstood your earlier comment. Removed them.

leighman commented:
Cheers!
